metaflow/client/cache/cache_client.py

```diff
@@ -0,0 +1,230 @@
+                    except:
+                        err = "Corrupted message: '%s'" % tail + buf
+                        raise CacheStreamCorrupted(err)
+                    else:
+                        tail = ''
+                        yield msg
+                else:
+                    tail += buf
+
+        if self.stream_key:
+            # the stream has three layers:
+            # 1) _readlines() reads raw JSON events from a file
+            # 2) action.stream_response() reformats the raw events into
+            #    action-specific output. Note that (2) may produce more
+            #    or fewer events than (1).
+            # 3) client.wait_iter() handles sync/async sleeping when no
+            #    events are available.
+            it = _readlines([self.stream_path, self.key_paths[self.stream_key]])
+            return self.client.wait_iter(self.action.stream_response(it),
+                                         timeout)
+
+class CacheClient(object):
+
+    def __init__(self, root, action_classes, max_actions=16, max_size=10000):
+
+        action_classes.append(Check)
+        for cls in action_classes:
+            setattr(self, cls.__name__, self._action(cls))
+
+        self._root = root
+        self._prev_is_alive = 0
+        self._is_alive = True
+        self._action_classes = action_classes
+        self._max_actions = max_actions
+        self._max_size = max_size
+
+    def start(self):
+        cmd, env = subprocess_cmd_and_env('cache_server')
+        cmdline = cmd + [\
+                   '--root', os.path.abspath(self._root),\
+                   '--max-actions', str(self._max_actions),\
+                   '--max-size', str(self._max_size)\
+                ]
+
+        msg = {
+            'actions': [[c.__module__, c.__name__] for c in self._action_classes]
+        }
+        return self.request_and_return([self.start_server(cmdline, env),
+                                        self._send('init', message=msg),
+                                        self.check()],
+                                       None)
+
+    def stop(self):
+        return self.stop_server()
+
+    @property
+    def is_alive(self):
+        return self._is_alive
+
+    def ping(self):
+        return self._send('ping')
+
+    def _send(self, op, **kwargs):
+        req = server_request(op, **kwargs)
+        return self.send_request(json.dumps(req).encode('utf-8') + b'\n')
+
+    def _action(self, cls):
+
+        def _call(*args, **kwargs):
+            msg, keys, stream_key, disposable_keys =\
+                 cls.format_request(*args, **kwargs)
+            future = CacheFuture(keys, stream_key, self, cls, self._root)
+            if future.is_ready():
+                # cache hit
+                req = None
+            else:
+                # cache miss
+                action_spec = '%s.%s' % (cls.__module__, cls.__name__)
+                req = self._send('action',
```

Comment:
the fire-and-forget nature of executing actions is problematic for error catching at the caller level. It would be preferable that something like this could be performed when calling actions:

```
try:
  await cache.SomeAction(params)
except:
  # any exception worth catching, CacheFullException, CacheServerUnreachable, etc.
  # Including the ability to raise an exception from within an actions execute()
```